File: python/03_s3_object_list_to_slack.py
import boto3
import json
import os
import urllib.request

def lambda_handler(event, context):
    client = boto3.client('s3')
    bucket = event['Records'][0]['s3']['bucket']['name']
    objects = client.list_objects_v2(Bucket=bucket)
    notify(', '.join(map(lambda x: x['Key'], objects['Contents'])))
    return {
        'statusCode': 200,
        'body': json.dumps('Hello from Lambda!')
    }

def notify(text):
    data = json.dumps({'text': text}).encode('utf-8')
    headers = {'Content-Type': 'application/json'}
    req = urllib.request.Request(
        os.environ['WEBHOOK_URL'],
        data=data,
        method='POST',
        headers=headers)
    with urllib.request.urlopen(req) as res:
        print(res.read().decode('utf-8'))

File: node/03_s3_object_list_to_slack.js
const https = require('https')
const AWS = require('aws-sdk')

exports.handler = async (event) => {
    const client = new AWS.S3()
    const bucket = event.Records[0].s3.bucket.name
    const res = await client.listObjectsV2({ Bucket: bucket }).promise()
    await notify(res.Contents.map(obj => obj.Key).join(', '))
    const response = {
        statusCode: 200,
        body: JSON.stringify('Hello from Lambda!'),
    }
    return response
}

const notify = text => {
    return new Promise((resolve, reject) => {
        const data = JSON.stringify({ text: text })
        const options = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            }
        }
        const req = https.request(process.env['WEBHOOK_URL'], options, res => {
            console.log(res.statusCode)
            resolve()
        })
        req.write(data)
        req.end()
    })
}
